### PR TITLE
fix source map for parameters, re-enable disassembler use of source map

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Fix disassembler use of move-compiler-v2 source map by fixing the way parameter types are added to the source map.  Re-enable the source map in disassembler.

The problem was that FunctionGenerator::temp_to_local (in move-compiler-v2/src/file_format_generator/function_generator.rs) is used to allocate both parameters and locals, but as a side-effect it calls self.gen.source_map.add_local_mapping(...), which should only be called for locals, as parameters are recorded elsewhere. This PR adds a flag to temp_to_local() to change the behavior in the case where it is used to allocate parameters fixes the problem.

Note that without this fix to function_generator.rs (but with the change to model.rs to re-enable source map in disassembly), then running
```
MOVE_COMPILER_V2=true cargo test -p move-cli
```
shows failed tests with errors like:
```
Command `disassemble --package Test --name m`:
Error: Unable to get type for local at index 0
```

Notably in these test outputs:
- third_party/move/tools/move-cli/tests/build_tests/build_with_warnings/args.exp
- third_party/move/tools/move-cli/tests/sandbox_tests/package_basics/args.exp  (twice)

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Other (specify)
   move-compiler-v2, disassembler
## How Has This Been Tested?

Ran various tests with MOVE_COMPILER_V2 enabled, in combination with PR#12943.

## Key Areas to Review

Need to identify the use case which actually triggered bug #12541 and test it with this.  Notably, `MOVE_COMPILER_V2=true cargo test -p move-cli` now shows reasonable disassembly outputs.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by rahxephon89 — APPROVED** [review]
The code looks good to me. Maybe you could explicitly mention in the PR description that parameters should not be added to the local map so this change is made?

File: third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs:180]: There seems to be useful discussion here about why we don't want to add to source maps here: https://github.com/aptos-labs/aptos-core/issues/12541#issuecomment-2053926508

We should provide some comments here as well explaining the newly added boolean and why it is used the way it is.

**Review by vineethk — COMMENTED** [review]
Could we please add some automated tests?

**Comment by brmataptos** [comment]
> Could we please add some automated tests?

Any suggestions about how to do that?  It shows up when using V2 (through environment variables) to run certain test cases.

**Comment by brmataptos** [comment]
Ok, I've realized I don't need an extra parameter to `temp_to_local` to figure out if the temp is a parameter, so code change is simplified.

And I've added some mechanism to move-cli to allow recording test results for V2 runs (as `.v2_exp` files).  If you want to see the output *before* the change, I've checked that in as a separate commit, and with the original fix I checked in the changes .v2_exp file outputs.

File: third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp
Comment (vineethk) [inline third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp:5]: I can see some weird chars in this exp file (but not in the corresponding v1 exp file)? Could you please double check these? Asking because I have not seen these in other exp files so far.

**Review by vineethk — COMMENTED** [review]
LGTM, but have a comment about weird chars in newly added v2 exp files. Did not want this PR to get automatically merged before that is looked into - can approve after that comment is addressed.

File: third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp
Comment (vineethk) [reply third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp:5]: Screenshot 2024-04-23 at 9 55 50 AM

A screenshot of what I am seeing in github.dev if that is helpful.

Comment (brmataptos) [reply third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp:5]: Those are ANSI color codes.  It's useful for humans, but should we not record it in test output files?

Comment (brmataptos) [reply third_party/move/tools/move-cli/tests/build_tests/dependency_chain/args.v2_exp]: Ok, figured out that it was already being done for move-compiler, added a new mechanism for others.

**Review by vineethk — APPROVED** [review]

**Pull request merged**
